### PR TITLE
APIVersion, Multi-page result fix and extension to Get-GitLabProjectCommit

### PR DESCRIPTION
**Pull request by markwragg**
- Added -APIVersion parameter to Save-GitLabAPIConfiguration  to allow other API versions to be defined/stored (was previously hard coded as v3).
- Amended QueryGitLabAPI and Test-GitLabAPI to make use of new APIVersion parameter.
- Modified Get-GitLabProjectCommit to include -Before, -After and -Branch parameters for filtering results via the API
- Fixed QueryGitLabAPI pagination issue, where it was using a reference object for $Request and so was not handling the query string for multiple pages correctly

I have run .\build.ps1 -Task Pre-Commit and there were no errors.

The help documentation needs to be updated regarding the new parameters. I'm not familiar with the way you're doing help so am not certain how to do this.

### CODE REVIEW

File: PSGitLab/Private/QueryGitLabAPI.ps1 [hunk 1]
```diff
@@ -21,6 +21,9 @@ param(
 )
 
 $GitLabConfig = ImportConfig
+
+if ($GitLabConfig.APIVersion) { $Version = "v$($GitLabConfig.APIVersion)" }
```
Comment (markwragg) [inline PSGitLab/Private/QueryGitLabAPI.ps1:25]: This approach is to make this a non-breaking change for anyone that hasn't re-run Save-GitLabAPIConfiguration where the APIVersion property now gets set.

File: PSGitLab/Private/QueryGitLabAPI.ps1 [hunk 2]
```diff
@@ -41,8 +44,8 @@ try  {
     $totalPages = ($webContent).Headers['X-Total-Pages'] -as [int]
     $Results = $webContent.Content | ConvertFrom-Json
     for ($i=1; $i -lt $totalPages; $i++) {
-        $newRequest = $Request
-        $newRequest.URI = $newRequest.URI + "&page=$($i+1)"
+        $newRequest = $Request.PSObject.Copy()
```
Comment (markwragg) [inline PSGitLab/Private/QueryGitLabAPI.ps1:47]: Prior to this in my environment (Win 10, PS5.1) $newrequest was acting like a reference to $request, so the "&pages" setting was duplicated on each iteration.

File: PSGitLab/Private/QueryGitLabAPI.ps1 [hunk 3]
```diff
@@ -41,8 +44,8 @@ try  {
...
+        $newRequest.URI = $newRequest.URI + "?&page=$($i+1)"
```
Comment (markwragg) [inline PSGitLab/Private/QueryGitLabAPI.ps1:48]: I also found that without the `?` prior to the querystring this was ending up as an invalid URL.

File: PSGitLab/Public/Projects/Get-GitLabProjectCommit.ps1 [hunk 4]
```diff
@@ -11,7 +11,18 @@
     [ValidateNotNull()]
     [ValidateNotNullOrEmpty()]
     #[Parameter(ParameterSetName='Namespace')]
```
Comment (markwragg) [inline PSGitLab/Public/Projects/Get-GitLabProjectCommit.ps1:13]: Should this be commented out? I have not modified this but it seems wrong.

File: PSGitLab/Public/Projects/Get-GitLabProjectCommit.ps1 [hunk 5]
```diff
@@ -26,9 +37,27 @@
     }
   }
 
+  $Body = @{}
```
Comment (markwragg) [inline PSGitLab/Public/Projects/Get-GitLabProjectCommit.ps1:40]: `Invoke-WebRequest` automatically turns the hashtable passed as -Body in to additional querystring parameters. This seemed the most logical way to pass these optional parameters.

**Comment by ngetchell** [comment]
This is some great work. Give me a little time to unpack it. 

I like how you solved the transition from v3-v4 of the API. Might still hard-code any APIs that don't work yet for v4. Or fix them is issue itself.

**Comment by markwragg** [comment]
Yeah I did very little testing of the API change hence keeping it v3 by default unless the user overrides. I had to add it for my need as my Server didn’t seem to return the pagination
header on v3.

**Comment by ngetchell** [comment]
Thanks for the contribution. Elegant way to handle the api version change.

**Pull request merged**